url8af84d296f_mrcosta_gantt_generator_tgz-pJ8-org_gantt_generator_mrcpsp_RunJ8/src/org/swiftgantt/model/Task.java

```diff
@@ -0,0 +1,622 @@
+		int ret = 0;
+		if (this.treeModel == null || this.getActualEnd() == null) {
+			return ret;
+		}
+		Config cfg = GanttChart.getStaticConfig();
+		TimeUnit timeUnit = this.treeModel.getTimeUnit();
+		Time et = (Time) this.getActualEnd().clone();
+		if (timeUnit != null && this.getActualStart() != null && et != null) {
+			if (timeUnit == TimeUnit.Hour) {
+				ret = CalendarUtils.calcWorkingHours(this.getActualStart(), et, cfg.getWorkingHoursOfDay(),
+						cfg.getWorkingHoursSpanOfDay()[0]);
+
+			} else if (timeUnit == TimeUnit.AllDay) {
+//				ret = CalendarUtils.getHourInterval(start, end) + 1;
+					ret = start.getHourIntervalTo(end) + 1;
+			} else if (timeUnit == TimeUnit.Day) {
+				ret = CalendarUtils.calcWorkingDays(this.getActualStart(), et, GanttChart.getStaticConfig().getWorkingDaysSpanOfWeek());
+
+			} else if (timeUnit == TimeUnit.Week) {
+				int yInterval = this.getActualEnd().getYear() - this.getActualStart().getYear() - 1;
+				int wInterval = (52 - this.getActualStart().get(Time.WEEK_OF_YEAR)) + this.getActualEnd().get(Time.WEEK_OF_YEAR);
+				ret = yInterval * 52 + wInterval;
+
+			} else if (timeUnit == TimeUnit.Month) {
+				int yInterval = this.getActualEnd().getYear() - this.getActualStart().getYear();
+				int mInterval = this.getActualEnd().getMonth() - this.getActualStart().getMonth() + 1;
+				ret = yInterval * 12 + mInterval;
+
+			} else if (timeUnit == TimeUnit.Year) {
+				ret = this.getActualEnd().getYear() - this.getActualStart().getYear() + 1;
+			}
+		//			logger.debug(" Duration for: " + this.name + " is: " + duration
+		//					+ " from: " + this.start.getTime() + " to: " + et.getTime() + " in time unit: " + timeUnit);
+		}
+		return ret;
+	}
+
+	/*
+	 * Check if start time of task is in rest-out time, adjust it to working time.
+	 */
+	private Time checkAndLimitStartTime(Time startTime) {
+		if (startTime == null || this.treeModel == null || this.treeModel.getTimeUnit() == null) {
+			return startTime;
+		}
+		Time adjustedStartTime = startTime.clone();
+		TimeUnit timeUnit = this.treeModel.getTimeUnit();
+
+		// Do nothing for other time unit.
+		if (timeUnit == TimeUnit.Hour) {
+			int lowBound = GanttChart.getStaticConfig().getWorkingHoursSpanOfDay()[0];
+			int highBound = lowBound + BaseTimeAxis.STEPS_OF_MAJOR_SCALE - 1;
+			int hourOfDay = startTime.get(Time.HOUR_OF_DAY);
+			if (hourOfDay < lowBound || hourOfDay > highBound) {
+				adjustedStartTime.set(Time.HOUR_OF_DAY, lowBound);
+			}
+			if (hourOfDay > highBound) {
+				adjustedStartTime.add(Time.DATE, 1);
+			}
+		} else if (timeUnit == TimeUnit.Day) {
+			int offset = 0;
+			int[] wds = GanttChart.getStaticConfig().getWorkingDaysSpanOfWeek();//Working day span
+			int dayOfWeek = startTime.get(Time.DAY_OF_WEEK);
+			if (dayOfWeek < wds[0]) {
+				offset = wds[0] - dayOfWeek;
+			} else if (dayOfWeek > wds[1]) {
+				offset = 7 - dayOfWeek + wds[0];
+			}
+			if (offset != 0) {
+				adjustedStartTime.add(Time.DATE, offset);
+			}
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 525:537, 560:572.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.